「Railsでテストを書く」の提出物

yocaji opened the pull request:
以下のテストを作成しました。お手数ですがご確認よろしくお願い致します🙇‍♂️

- モデルテスト
  - User
  - Report
- システムテスト
  - 日報の作成・編集・削除
  - 本の作成・編集・削除

### FactoryBotに関する疑問

`create(:foo)`する前に`FactoryBot.reload`していますが、これが無い場合は下記のようにエラーとなってしまいます。
ですが毎回reloadするのも妙な感じなので、この対応で良いのか疑問を持っています。

```console
UserTest#test_#name_or_email:
KeyError: Factory not registered: "alice"
    test/models/user_test.rb:8:in `block in <class:UserTest>'
```

Commit:
ハンズオン動画に沿ってUserのモデルテストを実装した

Commit:
Userモデルテストの残りを作成した

Commit:
Reportのモデルテストを作成した

Commit:
ハンズオン動画を参考に本のシステムテストを作成した

Commit:
不要なファイルを削除した

Commit:
日報のユニットテストを作成した

Commit:
モデルテストにFactoryBotを導入した

Commit:
システムテストにFactoryBotを導入した

Commit:
使わなくなったfixtureを削除した

Commit:
システムテストにおけるログイン処理をWardenを使うように変更した

JunichiIto reviewed (commented):
動画でレビューしたので別途リンクを送ります〜。

Commit:
5/1のレビューでいただいたコメントに基づいて修正した

- gemをfactory_bot_railsに変更した
- ファクトリファイルには1組のデータだけ記述するようにした
- 日報の投稿日付のテスト方法を修正した
- テストで比較するデータはテストファイル側に記述するようにした
- 使わないインスタンス変数を作成しないように修正した

yocaji commented on test/factories/books.rb:
```diff
@@ -0,0 +1,9 @@
+  factory :book do
+    title { '南総里見八犬伝' }
+    memo { '伝奇小説' }
+    author { '曲亭馬琴' }
```
ここで入れているデータは使わなくなったので、`nil`にするべきなのか迷ってひとまずそのままにしました。

JunichiIto commented on test/models/user_test.rb:
```diff
@@ -4,9 +4,8 @@
-    @alice = create(:alice)
-    @bob = create(:bob)
+    @alice = create(:user, name: 'アリス')
+    @bob = create(:user, email: 'bob@example.com')
```
`name: 'アリス'`と`email: 'bob@example.com'`は`test '#name_or_email'`のために設定しているんだと思いますが、テストデータとテストの実行位置が離れてしまうと、第三者がテストコードを読んだときに「なんでこうなってるの？」となってしまいがちなので、`test '#name_or_email'`の中で名前をセットしたり、名前をnilにしたり切り替えるのも一手です。

Everyday Railsでもどこかに「テストコードを読むときにスクロールが発生するようなテストはダメ」みたいな話が書いてあったと思います〜。

yocaji replied:
> テストコードを読むときにスクロールが...

確かにそうでした💦
教えたいただいたように、`@alice` の属性を書き換える方法に修正しました。

JunichiIto reviewed (commented):
コメントしました〜。

Commit:
#name_or_emailのためのテストデータの入れ方を修正した

JunichiIto reviewed (approved):
確認しました。OKです 🙆‍♂️
